### PR TITLE
PopSQL: add psalm, gh actions, param/return types

### LINKED ISSUE #4: Ordering is backwards in assertSame() and assertEquals()
[The PHPUnit signature](http://phpunit.de/manual/3.7/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.assertions.assertSame) for `assertSame()` is as follows:

```
assertSame(mixed $expected, mixed $actual[, string $message = ''])
```

In the test file, however, we repeatedly switch the order of the first two arguments:

```
$this->assertSame($actualQuery, $expectedQuery);
```

While this will of course work, it will cause a failing test to be a bit more confusing.

### LINKED ISSUE #15: Types: Add return types and arg types
Right now, we have no types so psalm doesn't know that `$q->select()` returns `$this` and thus thinks everything is mixed.

### Do This
* Add psalm type checking for CI
* Add types

### PR DESCRIPTION
**Pull request by davidrans**
Closes https://github.com/iFixit/PopSQL/issues/15
Closes https://github.com/iFixit/PopSQL/issues/4

### CODE REVIEW

File: QueryGenerator.php [hunk 1]
```diff
@@ -380,29 +395,37 @@ private function assertCompleteQuery() {
...
+   private static function getPrimaryClauses(): array {
       return array_keys(self::$possibleClauses);
    }
 
    /**
     * Return the primary clause in this QueryGenerator instance.
     * If multiple primary clauses have been set, all but the first set clause
     * will be ignored.
+    *
+    * @return string|false
```
Comment (djmetzle) [inline QueryGenerator.php:416]: This isn't necessary. `false` is a literal type now (as of 8.0):
https://php.watch/versions/8.0/union-types#false

File: QueryGenerator.php [hunk 2]
```diff
@@ -441,8 +464,10 @@ private function constructClause($method, $skipClause = false) {
    /**
     * return the appropriate glue string for the given clause, taking into
     * account $this->useOr
+    *
+    * @return string|false
```
Comment (djmetzle) [inline QueryGenerator.php:468]: Same as above.

**Review by djmetzle — DISMISSED** [review]
Looks good. CR 🎈

Just to ask, should we think about inlining this into the monorepo? It could save a headache with integration and action?

**Comment by davidrans** [comment]
> should we think about inlining this into the monorepo?

Maybe so. I think the only argument for keeping it separate is that it might be useful to the open source community, but that's probably far-fetched

**Comment by djmetzle** [comment]
@davidrans sure. We dont _have_ to archive it. And even if we did, it would still be publicly available.

Its not widely used:
image
(and id bet a dollar that at least one of those forks is an iFixiter)

**Review by djmetzle — APPROVED** [review]
CR :ginger_root:

**Pull request merged**